### PR TITLE
sql: Update sql package godoc

### PR DESCRIPTION
**Pull request by rjnn**
I've made a pass over the sql godoc, updating it where it was unclear to me. There are a couple TODOs remaining where I don't understand what the documentation is trying to say, and I'd appreciate some input/clarification to help expand/clarify those bits.

cc @paperstreet @andreimatei @knz

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/10706)

### CODE REVIEW

**Comment by RaduBerinde** [comment]
:lgtm:

---

Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/sql/doc.go, line 81 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10706#-KWe4-DE7DlfK9zhTKuQ:-KWe4-DE7DlfK9zhTKuR:bfbt28u) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f4bb1cc9c9350219675e274a401de678f549c25/pkg/sql/doc.go#L81)):*
> ```Go
> system runs the following query:
> 
>   SELECT id FROM system.namespace WHERE parentID = 0 AND name = <database-name>
> ```

NYC but it's a bit odd that we're using `system.namespace` for these examples.

---

*[pkg/sql/doc.go, line 127 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10706#-KWe1rxM3Ch6EahsHSjB:-KWe1rxM3Ch6EahsHSjC:b-f53jur) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f4bb1cc9c9350219675e274a401de678f549c25/pkg/sql/doc.go#L127)):*
> ```Go
> to detect rows which only contain NULL values in non-primary key columns, every
> row has a sentinel key indicating its existence. The sentinel key is simply the
> primary index key: TODO(Arjun): I do not understand this paragraph, it needs
> ```

The best explanation of this is in the blog post (Null Column Values section): https://www.cockroachlabs.com/blog/sql-in-cockroachdb-mapping-table-data-to-key-value-storage/


But this doc is missing documentation on column families (https://www.cockroachlabs.com/blog/sql-cockroachdb-column-families/), maybe @paperstreet wants to add something.

---

*[pkg/sql/doc.go, line 152 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10706#-KWe2dNYM0yRq8TgNivW:-KWe2dNYM0yRq8TgNivX:b-bunbp5) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f4bb1cc9c9350219675e274a401de678f549c25/pkg/sql/doc.go#L152)):*
> ```Go
> Notice that while the index is on "name", the key contains both "name" and
> "parentID". This is done to ensure that each row for a table has a unique key
> for the non-unique index (and therefore, a conditional put operation on the
> ```

The addition is correct but doesn't make sense in this sentence - the problem we are addressing by adding the PK to the key is that we can't have multiple values for the same key, so we're making the keys unique.

---

*[pkg/sql/doc.go, line 167 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10706#-KWe34T09yXvw7TkDoaN:-KWe34T09yXvw7TkDoaO:b-4wazkc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f4bb1cc9c9350219675e274a401de678f549c25/pkg/sql/doc.go#L167)):*
> ```Go
> insertion of a row into the table that would result in a duplication in the
> index will fail. However, we still need to be able to decode the full primary
> key by reading this index, for example, in order to efficiently execute the
> ```

Adding PKs is not just to make certain queries more efficient (you could make the same argument for any columns). We need the PK so that we can cross-reference the primary index: if we need any other columns, we look up the full row in the primary index by PK (index join).

---

*[pkg/sql/doc.go, line 190 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10706#-KWe453YIHzVtEPmpUCC:-KWe453YIHzVtEPmpUCD:b-pza4sn) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f4bb1cc9c9350219675e274a401de678f549c25/pkg/sql/doc.go#L190)):*
> ```Go
> Looking this up by the index on "name" does not give us the value of the "bar"
> column. Instead, to process this query, Cockroach does two key-value reads,
> which are morally equivalent to the following SQL:
> ```

from this sentence I expected a single SQL, e.g. `SELECT bar FROM system.namespace WHERE (parentID, name) IN (SELECT (parentID, name) FROM system.namespace WHERE name = "foo")`

---

*[pkg/sql/doc.go, line 192 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10706#-KWe4Y8xulN-N4NXtQqi:-KWe4Y8xulN-N4NXtQqj:bnqz12s) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f4bb1cc9c9350219675e274a401de678f549c25/pkg/sql/doc.go#L192)):*
> ```Go
> which are morally equivalent to the following SQL:
> 
>   SELECT (parentID, name) FROM systemnamespace WHERE "name = "foo";
> ```

system.namespace; extra " at "name

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10706#-:-KWe5c7lf6qq1EsK4eBR:bnfp4nl)*

File: pkg/sql/doc.go [hunk 1]
```diff
@@ -72,37 +74,40 @@ database/table name to ID and from ID to descriptor:
...
+integers, floating point numbers and strings such that the lexicographic
+ordering of the encoded strings corresponds to the same ordering of the
+unencoded data. Using `system.namespace` as an example, the primary key columns
+would be encoded as:
 
   /parentID/name
 
 [Note that "/" is being used to disambiguate the components of the key. The
-actual encodings do not use "/"].
+actual encodings do not use the "/" character].
```
Comment (knz) [inline pkg/sql/doc.go]: Perhaps add a reference here to where the encoding functions live.

File: pkg/sql/doc.go [hunk 2]
```diff
@@ -111,16 +116,16 @@ corresponds to the primary index:
...
+values for every column is stored at the key:
 
   /TableID/PrimaryIndexID/parentID/name/ColumnID -> Value
 
 A column containing a NULL value is not stored in the monolithic map. In order
 to detect rows which only contain NULL values in non-primary key columns, every
 row has a sentinel key indicating its existence. The sentinel key is simply the
-primary index key:
+primary index key: TODO(Arjun): I do not understand this paragraph, it needs
+clarification.
```
Comment (knz) [inline pkg/sql/doc.go]: Try this:

"""
A column containing a NULL value is simply omitted from the monolithic map; its absence indicates the NULL value.
To cover the case where all the non-PK columns of a row happen to be NULL, a special encoding is used: a "sentinel" key, which is simply the primary index key of that row without any column suffix, is used to indicate the presence of the row.
(FIXME: is the following true?) Sentinel keys are only used for tables that have zero non-null, non-PK columns.
"""

File: pkg/sql/doc.go [hunk 3]
```diff
@@ -129,116 +134,163 @@ separately as their data can be decoded from the sentinel value.
...
 Notice that while the index is on "name", the key contains both "name" and
 "parentID". This is done to ensure that each row for a table has a unique key
-for the non-unique index. In general, for a non-unique index we encoded the
+for the non-unique index (and therefore, a conditional put operation on the
+underlying key-value store is guaranteed to fail if the row already exists). In
+general, in order to guarantee that a non-unique index is unique, we encode the
 index's columns followed by any primary key columns that have not already been
-mentioned. This effectively transforms any non-unique index into a unique
-index.
+mentioned. Since the primary key must be uniquely define a row, this transforms
```
Comment (knz) [inline pkg/sql/doc.go]: "must be uniquely define a row" -> grammar

File: pkg/sql/doc.go [hunk 4]
```diff
@@ -129,116 +134,163 @@ separately as their data can be decoded from the sentinel value.
...
-how well they match the sorting required by the ORDER BY clause. Back to the
-example above, the range information would determine that:
+index to scan? We currently use a restricted form of value propagation
+(TODO(arjun): Elaborate, this is difficult to understand) in order to determine
+the range of possible values for columns referenced in the WHERE clause. Using
+this range information, each index is examined to determine if it is a potential
+candidate and ranked according to its specificity. In addition to ranking
+indexes by the column value range information, they are ranked by how well they
+match the sorting required by the ORDER BY clause. Back to the example above,
+the range information would determine that:
```
Comment (knz) [inline pkg/sql/doc.go]: Maybe around here add link to Radu's blog post about index selection.

**Comment by arjunravinarayan** [comment]
TFTRs, @raduberinde and @knz, some very valuable feedback there. I'll incorporate and push soon.

---

Review status: 0 of 1 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/doc.go, line 81 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10706#-KWe4-DE7DlfK9zhTKuQ:-KWiNeo0mcz9JtinnJYv:bmry6yu) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f4bb1cc9c9350219675e274a401de678f549c25/pkg/sql/doc.go#L81)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> NYC but it's a bit odd that we're using `system.namespace` for these examples.
</details>
Yeah, good point, I'll try rewriting it with a different `example` table.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10706#-:-KWixaBfMEZCCAa8mX2a:b-y33wlm)*

**Comment by arjunravinarayan** [comment]
Updated, with a different example, and removing all TODOs.

---

Review status: 0 of 1 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/sql/doc.go, line 110 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10706#-KWdaVIkfxZbZmMTGZVK-r1-110:-KWizwzWcIIIqZgk2oZK:b-q4l1bd) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f4bb1cc9c9350219675e274a401de678f549c25/pkg/sql/doc.go#L110)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Perhaps add a reference here to where the encoding functions live.
</details>
A few lines above I say `(the encoding is specified in the `util` package in `util/encoding).` Would you like me to say something more?

---

*[pkg/sql/doc.go, line 127 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10706#-KWe1rxM3Ch6EahsHSjB:-KX7xyidyez1wvyDeBYI:bjfy9rg) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f4bb1cc9c9350219675e274a401de678f549c25/pkg/sql/doc.go#L127)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> The best explanation of this is in the blog post (Null Column Values section): https://www.cockroachlabs.com/blog/sql-in-cockroachdb-mapping-table-data-to-key-value-storage/
> 
> 
> But this doc is missing documentation on column families (https://www.cockroachlabs.com/blog/sql-cockroachdb-column-families/), maybe @paperstreet wants to add something.
</details>
Excellent. Thank you! I have added a section on column values. PTAL.

---

*[pkg/sql/doc.go, line 128 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10706#-KWdaVIkfxZbZmMTGZVK-r1-128:-KX7ySSiZJd5UjDhntF1:bhcu64y) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f4bb1cc9c9350219675e274a401de678f549c25/pkg/sql/doc.go#L128)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Try this:
> 
> 
> 
> """
> 
> A column containing a NULL value is simply omitted from the monolithic map; its absence indicates the NULL value.
> 
> To cover the case where all the non-PK columns of a row happen to be NULL, a special encoding is used: a "sentinel" key, which is simply the primary index key of that row without any column suffix, is used to indicate the presence of the row.
> 
> (FIXME: is the following true?) Sentinel keys are only used for tables that have zero non-null, non-PK columns.
> 
> """
</details>
I believe that statement is false, but have very little confidence in my answer. I am basing off of the blog post that says "CockroachDB always writes a sentinel key for each row that is the primary key without a column ID suffix."

---

*[pkg/sql/doc.go, line 152 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10706#-KWe2dNYM0yRq8TgNivW:-KX7oyMO9LGnRx2NfKar:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f4bb1cc9c9350219675e274a401de678f549c25/pkg/sql/doc.go#L152)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> The addition is correct but doesn't make sense in this sentence - the problem we are addressing by adding the PK to the key is that we can't have multiple values for the same key, so we're making the keys unique.
</details>
Done.

---

*[pkg/sql/doc.go, line 156 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10706#-KWdaVIkfxZbZmMTGZVK-r1-156:-KX7p0nWqg1f8L1VCCbx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f4bb1cc9c9350219675e274a401de678f549c25/pkg/sql/doc.go#L156)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> "must be uniquely define a row" -> grammar
</details>
Done.

---

*[pkg/sql/doc.go, line 167 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10706#-KWe34T09yXvw7TkDoaN:-KX7pCs9E6tgay2194_2:b4h84hj) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f4bb1cc9c9350219675e274a401de678f549c25/pkg/sql/doc.go#L167)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Adding PKs is not just to make certain queries more efficient (you could make the same argument for any columns). We need the PK so that we can cross-reference the primary index: if we need any other columns, we look up the full row in the primary index by PK (index join).
</details>
Yes, this is what I meant to say, I've refactored. Let me know if its not a sufficient improvement.

---

*[pkg/sql/doc.go, line 190 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10706#-KWe453YIHzVtEPmpUCC:-KX7ukw9pfyMIvRQCf1-:b-nq1wdm) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f4bb1cc9c9350219675e274a401de678f549c25/pkg/sql/doc.go#L190)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> from this sentence I expected a single SQL, e.g. `SELECT bar FROM system.namespace WHERE (parentID, name) IN (SELECT (parentID, name) FROM system.namespace WHERE name = "foo")`
</details>
Since the SQL notation is an analogy to explain the underlying key-value operations, I think it is simpler to keep them separate, so just changed the sentence, rather than the SQL.

---

*[pkg/sql/doc.go, line 192 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10706#-KWe4Y8xulN-N4NXtQqi:-KX7ukdqJqyr07zUCmwE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f4bb1cc9c9350219675e274a401de678f549c25/pkg/sql/doc.go#L192)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> system.namespace; extra " at "name
</details>
Done.

---

*[pkg/sql/doc.go, line 268 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10706#-KWdaVIkfxZbZmMTGZVK-r1-268:-KX7v8TjtVcP9EmvOcx-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f4bb1cc9c9350219675e274a401de678f549c25/pkg/sql/doc.go#L268)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Maybe around here add link to Radu's blog post about index selection.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10706#-:-KX86HWX4ZjvCA8VUpPf:b7rf8z2)*

**Comment by knz** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/sql/doc.go, line 110 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10706#-KWdaVIkfxZbZmMTGZVK-r1-110:-KX8NA9bx-vSF_qC3jQm:b-hay2ix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f4bb1cc9c9350219675e274a401de678f549c25/pkg/sql/doc.go#L110)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary>
> A few lines above I say `(the encoding is specified in the `util` package in `util/encoding).` Would you like me to say something more?
</details>
Thanks!

---

*[pkg/sql/doc.go, line 128 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10706#-KWdaVIkfxZbZmMTGZVK-r1-128:-KX8NQGBfYqSn186DbEN:b-hhqfbo) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f4bb1cc9c9350219675e274a401de678f549c25/pkg/sql/doc.go#L128)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary>
> I believe that statement is false, but have very little confidence in my answer. I am basing off of the blog post that says "CockroachDB always writes a sentinel key for each row that is the primary key without a column ID suffix."
</details>
Feels like room for optimization :)
Thanks for checking!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10706#-:-KX8Nf5bkP5jdTOwykAe:bnfp4nl)*

**Pull request merged**